broadinstitute/scp-ingest-pipeline
# Update AnnData file extensions (SCP-4557)

Anndata files (typically .h5ad) are also valid HDF5 files (.h5 and .hdf5). SCP has a few "other" files with .h5 suffix. While we expect most Anndata files to have the .h5ad suffix, ingest should try and see if Scanpy will successfully open files ending in .h5 and .hdf5.

This PR also updates the "Unsupported file format" error message that currently conveys expected file MIME type information. Instead we'll now return the expected file extensions which may be more actionable for our users.

Manual test:

activate the scp-ingest-pipeline repo virtualenv
then from the ingest directory of the scp-ingest-pipeline repo, perform this setup:

```
source ../scripts/setup_mongo_dev.sh <path/to/your/Github/token>
unset BARD_HOST_URL
```

Run the following ingest job to confirm that .h5 files ingest without error:
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_h5ad --ingest-h5ad --h5ad-file ../tests/data/h5ad/test.h5`

You'll see information about the Anndata object and the job status is 'success'

> AnnData object with n_obs × n_vars = 238 × 187 backed at '../tests/data/h5ad/test.h5'
>     obs: 'submission_id', 'biosample_id', 'biosample_type', 'library_preparation_protocol', 'library_preparation_protocol__ontology_label', 'donor_id', 'species', 'species__ontology_label', 'mouse_strain', 'mouse_strain__ontology_label', 'organ', 'organ__ontology_label', 'brain_region', 'brain_region__ontology_label', 'sex', 'organism_age', 'organism_age__unit', 'organism_age__unit__ontology_label', 'disease', 'disease__ontology_label', 'seurat_clusters', 'cell_type', 'cell_type__ontology_label'
> distinct_id: 2f30ec50-a04d-4d43-8fd1-b136a2045079
> studyAccession: SCPdev
> fileName: 5dd5ae25421aa910a723a337
> fileType: input_validation_bypassed
> fileSize: 1
> trigger: dev-mode
> logger: ingest-pipeline
> appId: single-cell-portal
> status: success


Run the following ingest job to confirm that .h5 files ingest without error:
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_h5ad --ingest-h5ad --h5ad-file ../tests/data/h5ad/bad.foo`

Confirm the file is rejected with an error message like:

> File type not detected for ../tests/data/h5ad/bad.foo, expected file endings are: .h5ad .h5 .hdf5

This PR supports SCP-4525 (frontend UI for h5ad ingest) and fulfills SCP-4557.
